HW3/P5/label_regions.cl

```diff
@@ -80,20 +83,54 @@ propagate_labels(__global __read_write int *labels,
     old_label = buffer[buf_y * buf_w + buf_x];
 
     // CODE FOR PARTS 2 and 4 HERE (part 4 will replace part 2)
+    if( lx == 4 && ly == 4 ) // let (4,4) be the one thread that fetches grandparents
+    {
+        // force thread (4,4) to fetch grandparents for every work-group pixel that requires it 
+        // (excluding background and repeat immediate fetches)
+        for( int y_index = 0; y_index < get_local_size(0); y_index++ )
+            for( int x_index = 0; x_index < get_local_size(0); x_index++ )
+                if( buffer[(y_index + halo) * buf_w + halo + x_index] != w*h )
+                    if( buffer[(y_index + halo) * buf_w + halo + x_index] != last_read )
```

Comment:
If the values are equal, you should update the buffer location with the grandparent value (of the last_read location).